comms/src/outbound_message_service/outbound_message.rs

```diff
@@ -0,0 +1,110 @@
+//
+//  Redistribution and use in source and binary forms, with or without modification, are permitted provided that the
+//  following conditions are met:
+//
+//  1. Redistributions of source code must retain the above copyright notice, this list of conditions and the following
+//  disclaimer.
+//
+//  2. Redistributions in binary form must reproduce the above copyright notice, this list of conditions and the
+//  following disclaimer in the documentation and/or other materials provided with the distribution.
+//
+//  3. Neither the name of the copyright holder nor the names of its contributors may be used to endorse or promote
+//  products derived from this software without specific prior written permission.
+//
+//  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES,
+//  INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+//  DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+//  SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+//  SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+//  WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE
+//  USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+use crate::{
+    connection::message::{Frame, MessageError},
+    peer_manager::node_id::NodeId,
+};
+use chrono::prelude::*;
+use rmp_serde;
+use serde::{de::DeserializeOwned, Deserialize, Serialize};
+use std::convert::TryFrom;
+
+#[derive(Clone, Debug, Deserialize, Serialize, PartialEq)]
+pub struct OutboundMessage<T> {
+    pub destination_node_id: NodeId,
+    pub retry_count: u32,
+    pub creation_timestamp: DateTime<Utc>,
+    pub last_retry_timestamp: Option<DateTime<Utc>>,
+    pub message_envelope: T,
+}
+
+impl<T: Serialize + DeserializeOwned> OutboundMessage<T> {
+    /// Create a new OutboundMessage from the destination_node_id and message_envelope
+    pub fn new(destination_node_id: NodeId, message_envelope: T) -> OutboundMessage<T> {
+        OutboundMessage {
+            destination_node_id,
+            retry_count: 0,
+            creation_timestamp: Utc::now(),
+            last_retry_timestamp: None,
+            message_envelope,
+        }
+    }
+
+    /// Serialize an OutboundMessage into a single frame
+    pub fn to_frame(&self) -> Result<Frame, MessageError> {
+        let mut buf: Vec<u8> = Vec::new();
+        match self.serialize(&mut rmp_serde::Serializer::new(&mut buf)) {
+            Ok(_) => Ok(buf.to_vec()),
+            Err(_) => Err(MessageError::SerializeFailed),
+        }
+    }
+
+    /// Update the retry count and retry timestamp after a transmission attempt
+    pub fn transmission_attempt(&mut self) {
```

Comment:
I'd maybe name this `mark_transmission_attempt`

Reply:
Change applied